application/controllers/UploaderController.php

```diff
@@ -237,11 +237,43 @@ public function run($actionID)
                 echo ls_json_encode($return);
                 Yii::app()->end();
             }
-
+            // event to allow updating some part
+            $event = new PluginEvent('beforeMovePublicFileUploaded');
+            $event->set('surveyId', $surveyid);
+            $event->set('responseId', Yii::app()->session['survey_' . $surveyid]['srid']); // NULL if not exist
+            $event->set('qid', $oQuestion->qid);
+            $event->set('success', true);
+            $event->set('filename', $filename);
+            $event->set('size', $size);
+            $event->set('tmp_name', $_FILES['uploadfile']['tmp_name']);
+            App()->getPluginManager()->dispatchEvent($event);
+            $success = $event->get('success', true);
+            $filename = $event->get('filename');
+            $uploadfile_tmp_name = $event->get('tmp_name');
+            if ($uploadfile_tmp_name !== $_FILES['uploadfile']['tmp_name']) {
+                @unlink($_FILES['uploadfile']['tmp_name']);
+            }
+            if (!$success) {
+                @unlink($uploadfile_tmp_name);
```

Comment:
I KNOW it allow deletion of any file : did we need to control ?

Reply:
In case of error, my thought:
- I believe the unlinking of $uploadfile_tmp_name should happen inside the plugin.
- Outside the plugin, we delete $_FILES['uploadfile']['tmp_name'] if still there.
What do you think?

Reply:
> I believe the unlinking of $uploadfile_tmp_name should happen inside the plugin.

Yes, then deletion of invalid files are done in plugin : :+1: